Add a Process class to control execution of subprocesses and communicate with them.

EIREXE opened the pull request:
Uses TinyProcessLibrary (https://gitlab.com/eidheim/tiny-process-library/), which takes care of dumb stuff like flatpak portals.

Fixes https://github.com/godotengine/godot-proposals/issues/216

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

EIREXE committed:
Add a Process class to control execution of subprocesses and communicate with them.

Uses TinyProcessLibrary (https://gitlab.com/eidheim/tiny-process-library/), which takes care of dumb stuff like flatpak portals.

Fixes godot-proposals#216a

bruvzg commented on thirdparty/tiny-process-library/LICENSE:
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
Third-party code should be listed in the:

https://github.com/godotengine/godot/blob/master/thirdparty/README.md
https://github.com/godotengine/godot/blob/master/COPYRIGHT.txt

bruvzg commented on godot-changes.patch:
```diff
@@ -0,0 +1,45 @@
+diff --git a/thirdparty/tiny-process-library/process_unix.cpp b/thirdparty/tiny-process-library/process_unix.cpp
```
This should be in the `/thirdparty/tiny-process-library/patches`, not here.

bruvzg commented on drivers/unix/process_unix.cpp:
```diff
@@ -0,0 +1,102 @@
+/*************************************************************************/
+/*  process_unix.cpp                                                     */
```
Since it's used on Windows as well, it's probably should be something generic, not `Unix`.

neikeq commented:
Does TinyProcessLibrary properly clean up to avoid leaving defunct processes? It should listen for SIGCHLD and call wait() or waitpid().

EIREXE commented:
> Does TinyProcessLibrary properly clean up to avoid leaving defunct processes? It should listen for SIGCHLD and call wait() or waitpid() (see #7774).

Is it necessary for tinyprocesslibrary to handle it since os_unix already handles this? (Unless I missed something)

neikeq commented:
The reason it's no longer handled by OS_Unix is because the implementation was a bad one. It was reaping all child processes, even the ones Godot didn't launch. This was wrong because it was messing with other libraries that launch processes. For example, it was reverted because it was causing issues with `pclose` but even before that it was reaping processes launched from C#, for example. This means those libraries wouldn't be notified whenever the process exited (unless they poll) and wouldn't be able to retrieve the exit code.

A proper implementation of a SIGCHLD handler should only reap processes it knows about and leave the rest to other handlers by chaining the handler. Precisely because of this, if we were to re-introduce this, this time with a proper implementation, it wouldn't reap any process launched by TinyProcessLibrary.

Godot worked around the need for a SIGCHLD handler by using `setsid()`. This way, Godot no longer needs to reap the process, but I don't know if it has any downside (Would it still be possible to retrieve stdout/stderr and check the exit code? Or is it only a fire and forget solution?). The same could be said about the [double fork trick](https://stackoverflow.com/a/23938054).

EIREXE commented:
I see, I will do some testing

EIREXE commented:
I have made all changes, I think it should be ok now.

Faless commented on core/SCsub:
```diff
@@ -97,6 +97,20 @@ thirdparty_minizip_sources = ["ioapi.c", "unzip.c", "zip.c"]
+thirdparty_tiny_process_library_dir = "#thirdparty/tiny-process-library/"
+thirdparty_tiny_process_library_sources = []
+
+if env["platform"] in ["linuxbsd", "macos", "web", "android"]:
```
I don't think the library do anything on the web, so we should exclude the third party library there I guess.

Faless commented:
Given that `process_tiny_process_lib` is an implementation of `process`, maybe we should moved it to drivers?

Faless commented:
@EIREXE I think you forgot to add the `SCsub` in the `driver/tiny-process-lib` and should contain more or less what is currently in the core `SCsub`. Something like (not tested sorry):

```python

Import("env")

# Tiny-process-library
thirdparty_tiny_process_library_dir = "#thirdparty/tiny-process-library/"
thirdparty_tiny_process_library_sources = []

if env["platform"] in ["linuxbsd", "macos", "android"]:
    thirdparty_tiny_process_library_sources = ["process.cpp", "process_unix.cpp"]
elif env["platform"] == "windows":
    thirdparty_tiny_process_library_sources = ["process.cpp", "process_win.cpp"]

thirdparty_tiny_process_library_sources = [
    thirdparty_tiny_process_library_dir + file for file in thirdparty_tiny_process_library_sources
]

env.add_source_files(env.drivers_sources, thirdparty_tiny_process_library_sources)
env.add_source_files(env.drivers_sources, "*.cpp")

```

Faless commented:
> That's for the third party directory, which as I understand has everything in the thirdparty scsub

thirdparty doesn't have it's own SCsub (that code is in this PRs `core/SCsub`) if you check [`drivers/png/SCsub`](https://github.com/godotengine/godot/blob/master/drivers/png/SCsub) each "driver" includes its own thirdparty library (usually conditionally on a `builtin_libname` flag).

akien-mga commented:
> each "driver" includes its own thirdparty library (usually conditionally on a `builtin_libname` flag).

Here the `builtin_*` option can likely be left out, it's only meant to be able to build against system provided packages on Linux and I don't think this one is typically packaged as a shared lib.

EIREXE commented:
Made some changes:

- Made adding and removing sigchld callbacks thread safe
- Fixed a crash when getting the process ID on ProcessUnix, not sure why it works but it works from my testing
- Added a auto quote arguments on windows like OS.execute does.

neikeq commented:
A few comments on the SIGCHLD handler:

- AFAIK, you shouldn't lock inside a signal handler.
- Third-party code may have their own SIGCHLD handler, so you should use chaining to not override theirs:
  ```cpp
  static struct sigaction old_sigaction;
  
  void init() {
  	sigaction(SIGCHLD, &sa, &old_sigaction);
  }
  
  void handle_sigchld(int sig, siginfo_t *info, void *ucontext) {
  	if (did_we_spawn_this(info)) {
  	 	// reap and whatever else
  	 	return;
  	}
  
  	// not handled (not a pid we created), pass it to the next signal handler
  	if (old_sigaction.sa_handler != SIG_IGN && old_sigaction.sa_handler != SIG_DFL)
  		old_sigaction.sa_handler(sig, info, ucontext);
  }
  ```

Check #8648 as an old attempt at implementing this.

EIREXE commented:
> AFAIK, you shouldn't lock inside a signal handler.

What do you suggest we do instead? the code may attempt to run the handler after the process instances have been removed from another thread

> Third-party code may have their own SIGCHLD handler, so you should use chaining to not override theirs:

Sure, I'll do that.

EIREXE commented:
Added sigchld chaining

silverkorn commented:
Since many attempts were made on different PRs by veterans and abandoned to be superseded by this one 
(https://github.com/godotengine/godot/pull/8310 <sup><sub>(2017)<sub/></sup> & https://github.com/godotengine/godot/pull/68733 <sup><sub>(2022)<sub/></sup>), is there still consideration about the implementation of a Process class?

Sorry, I know it sounds like a "bump" but I just want to make sure it's not just because it went out of the contributors' mind due to scattered old PRs about the subject.

---

This implementation should satisfy my comments on other threads related to this one:
- https://github.com/godotengine/godot-proposals/issues/216#issuecomment-730483067
- https://github.com/godotengine/godot/pull/68733#issuecomment-1729698294

Calinou commented:
> Edit: Is this implemented with this recent PR (https://github.com/godotengine/godot/pull/89206)?

I believe so. I suggest trying it in your own projects and reporting feedback :slightly_smiling_face:
